app/src/main/kotlin/com/wire/android/ui/connection/ConnectionActionButton.kt

```diff
@@ -0,0 +1,209 @@
+/*
+ * Wire
+ * Copyright (C) 2023 Wire Swiss GmbH
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+
+package com.wire.android.ui.connection
+
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.Spacer
+import androidx.compose.foundation.layout.height
+import androidx.compose.foundation.layout.padding
+import androidx.compose.material3.Icon
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.platform.LocalInspectionMode
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.core.os.bundleOf
+import com.sebaslogen.resaca.hilt.hiltViewModelScoped
+import com.wire.android.R
+import com.wire.android.model.ClickBlockParams
+import com.wire.android.model.LoadableState
+import com.wire.android.navigation.EXTRA_CONNECTION_STATE
+import com.wire.android.navigation.EXTRA_USER_ID
+import com.wire.android.navigation.EXTRA_USER_NAME
+import com.wire.android.ui.common.button.WireButtonState
+import com.wire.android.ui.common.button.WirePrimaryButton
+import com.wire.android.ui.common.button.WireSecondaryButton
+import com.wire.android.ui.common.dimensions
+import com.wire.android.ui.snackbar.LocalSnackbarHostState
+import com.wire.android.ui.snackbar.collectAndShowSnackbar
+import com.wire.kalium.logic.data.user.ConnectionState
+import com.wire.kalium.logic.data.user.UserId
+
+@Composable
+fun ConnectionActionButton(
+    userId: UserId,
+    userName: String,
+    connectionStatus: ConnectionState
+) {
+    val viewModel: ConnectionActionButtonBaseViewModel = if (LocalInspectionMode.current) {
+        ConnectionActionButtonPreviewModel(LoadableState(connectionStatus))
+    } else {
+        hiltViewModelScoped<ConnectionActionButtonViewModel>(
```

Comment:
Here the Composable needs to manage:

- BaseViewModel
- ViewModel
- PreviewModel

It's a lot of boilerplate code.

I'd prefer if we could generate this automatically some how. I'll give it some thought, as this is an improvement that can be done later.

On a similar note, what if we keep the naming like we already do with other instances:

```kotlin
interface FooViewModel
class FooViewModelImpl: FooViewModel
class FooViewModelPreview: FooViewModel
```

So the default `ConnectionActionButtonViewModel` is actually the interface that the composables can use, and the implementation is explicitly marked as such.